### ISSUE TITLE
[DFA-arch] --clip-grad on the lstm lane: the FAIR BPTT control for toy#157's stability claim (is BP's fragility BPTT's, or ours?)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#162]
## Why

toy#157 measured that **BPTT is the fragile arm** on the delayed-cue task, and Tao is recording that as a *stabiliser* mechanism for DFA alongside F17's regulariser. The measurement is real — at 4000 steps the per-step DFA cut solves 12 of 12 (lr × seed) cells while BPTT solves 7, and *which* rate works depends on the seed:

| lr | BP (seeds 0/1/2) | DFA step cut (seeds 0/1/2) |
|---|---|---|
| 0.005 | 0.504 / 1.000 / 0.996 | 1.000 / 1.000 / 1.000 |
| 0.010 | 0.250 / 1.000 / 0.996 | 1.000 / 1.000 / 1.000 |
| 0.020 | 0.250 / 0.996 / 1.000 | 1.000 / 0.996 / 1.000 |
| 0.030 | **1.000** / 0.250 / 0.238 | 1.000 / 1.000 / 0.996 |

**But toy has no gradient clipping, and clipping is the standard fix for exactly that failure mode** (Pascanu et al. 2013 — exploding gradients through long BPTT). Warmup was swept as the nearest available damper and it helps but is itself cell-dependent. So part of BP's fragility here is plausibly *ours*, not the method's, and the lane's own write-up says so:

> Some of BP's fragility is ours, not the method's. […] A clipped BPTT arm would be the fairer control — worth deciding before F21 leans on the stability half of this result.

This ticket builds that control. It is a cheap, self-contained de-risker for a claim that is otherwise about to be published.

## What to build

`--clip-grad NORM` on the `lstm` lane (env `LSTM_CLIP_GRAD`), off by default so **every existing configuration is byte-identical**:

```
toy train lstm --steps 4000 --lr 0.02 --clip-grad 1.0 --policy chain
```

- **GLOBAL-norm clipping**, the form the BPTT literature means: one norm over all stepped parameters, `scale = clip / max(clip, ||g||)`, applied to every gradient before the AdamW moments see it.
- It must be **in-graph and before the optimizer**. Scaling the learning rate instead is NOT equivalent under Adam — clipping changes what `m` and `v` accumulate, LR scaling does not. Buildable from existing ops (`sum`, `sqrt`, `div`, `scale_bias`, `mul` with a `[1,1]` broadcast); no new shim functions, so no CUDA/Metal mirror obligation.
- Only parameters that are actually stepped join the norm (a `frozen` layer contributes nothing), and the runner should say so in the bundle.

## What to measure

Re-run toy#157's fragility grid with clipping on, all three seeds:

1. **Does BP become uniformly reliable?** If yes, the stability half of toy#157 is substantially a harness artifact and both the roadmap and Tao's F21 conclusion must be corrected — that is the outcome this ticket exists to catch.
2. If BP stays seed-fragile *with* clipping, the claim gets stronger and can be stated without the caveat.
3. **The accuracy result should not move**: the per-step cut already solves the task at every cell, and clipping must not be what makes it work. Assert that the DFA arms are unchanged or better.

Also worth one cell: does clipping let BP reach the fair cell at a rate it previously failed at (lr 0.03, seeds 1/2)?

## Scope / discipline

- **Default off, byte-identical when absent** — the fixture and every published cell must not move ([[default-that-changes-nothing]], and toy#157 shipped a default that relabelled a consumer's experiment within hours; not repeating that).
- **lstm only.** The ssm lane's BP arm already reaches 1.000, so its comparison is not at risk; widening is a follow-on if this one finds something.
- CPU-only, as the rest of T0–T4.

## Gate legs (suggested)

- `--clip-grad` absent is BYTE-NULL against today's curve (the whole default-changes-nothing discipline).
- A very large clip norm is byte-null too (clipping that never binds must be a no-op), and a very small one is NOT (it must actually bind).
- Clipping is inert on a `frozen` arm (nothing is stepped).
- Determinism under clipping.
- The re-measured fragility row, whichever way it comes out, asserted so it cannot silently flip.

## What this does NOT do

It does not add clipping to the other lanes, and it does not re-open toy#157's accuracy result (the per-step cut matches BP at BP's own best cell on all three seeds — that stands either way). It answers exactly one question: **is BPTT's fragility on this task a property of BPTT, or of a harness that was missing its standard stabiliser?**

## Refs
- toy#157 (the fragility grid + the stated caveat), tao#20 (where the stabiliser mechanism is being recorded)
- Pascanu, Mikolov & Bengio 2013 — *On the difficulty of training recurrent neural networks* (norm clipping)
- toy#152 (DFA tolerates less LR — the neighbouring axis), [[default-that-changes-nothing]]

**Comment by OriPekelman** [comment]
## Shipped in `800add3` — the control was built, and toy#157's claim SURVIVED it

`--clip-grad NORM` (env `LSTM_CLIP_GRAD`): global-norm clipping over all stepped parameters, in-graph, **before** the AdamW moments. Off by default and byte-null when absent, so every cell this lane published still means what it said.

### It does not rescue BP

Cells SOLVED (val ≥ .9) out of the 12 (lr × seed) at 4000 steps:

| arm | unclipped | clip 1.0 | clip 0.1 |
|---|---|---|---|
| chain (BP) | 7/12 | **8/12** | **7/12** |
| dfa (step cut) | 12/12 | **12/12** | — |

### What it actually does is re-roll the dice

At clip 0.1, seed 0 solves at every rate while seed 2 mostly fails — the exact inverse of the unclipped pattern:

| cell | unclipped | clip 1.0 | clip 0.1 |
|---|---|---|---|
| lr .01, seed 0 | .250 | .250 | **1.000** |
| lr .03, seed 1 | .250 | **.992** | **1.000** |
| lr .005, seed 2 | .996 | 1.000 | **.266** |

**That non-monotonicity is the diagnosis.** If this were plain exploding gradients, a tighter clip would help monotonically. It doesn't: the arm is BIMODAL — it either catches the cue or it doesn't — and every hyperparameter, clipping included, just re-rolls which basin you land in. The per-step DFA cut is the only arm invariant to all of them (lr, warmup, clip, seed).

So the honest phrasing changes, and it is stronger than before: not *"BPTT explodes and toy was missing the fix"* but **"BPTT's success on this task is basin luck that no standard stabiliser removes, while cutting the time axis removes the basin problem entirely."** The roadmap caveat is deleted rather than reworded, and F21 can record the stabiliser mechanism.

### Implementation notes

- **Could not be done on the learning rate.** Under Adam, clipping changes what `m` and `v` *accumulate*; an LR scale does not. An hp-vector "clip" would have been a different experiment wearing this one's name.
- Built from existing ops (`sum`/`sqrt`/`div`/`scale_bias`/`mul` with a `[1,1]` broadcast), so **no new shim function** and no CUDA/Metal mirror obligation. `max(c, n)` is `(n + c + |n − c|)/2` with `|x| = sqrt(x·x)` — ggml has no max/abs.
- A `frozen` layer contributes nothing to the norm (clipping a gradient nobody applies would change the scale other layers see for no reason). The head is always stepped, so it always joins.
- `lstm` only. The ssm lane's BP arm already reaches 1.000 so its comparison was never at risk; `--clip-grad` is rejected on every other recipe rather than offering a knob whose effect is unmeasured there.

### Gated

- absent, and a 1e9 norm, are both **byte-null** against today's curve (a clip that cannot bind must be a no-op, or it is perturbing rather than clipping);
- a 0.01 norm **does** change it (a clip that never binds controls nothing);
- deterministic under clipping;
- **the finding itself**: at lr 0.01 / seed 0 / clip 1.0, BP reads .250 and the per-step cut reads 1.000. If clipping ever *did* rescue BP there, that leg fails loudly and names the correction — roadmap and tao#20 — rather than letting the claim rot.
- `--clip-grad 0` fails loud (omitting the flag is how you disable it; a 0 would be a setting that silently means "unset").

`GATE PASS [lstm]` (13 legs), `GATE PASS [train-cli-matrix]` (102 rejections + 12 acceptance probes).

**Comment by OriPekelman** [comment]
Closing: shipped in `800add3` — feat(dfa-arch): toy#162 --clip-grad on the lstm lane — the FAIR BPTT control, and toy#157's stability claim SURVIVES it

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
